### PR TITLE
feat: adds cilium v.1.8 support

### LINKED ISSUE #3290: current cilium implementation doesn't work
We would need to do some debugging to discover what's atrophied, or alternatively we just replace the existing implementation entirely with a more modern cilium implementation.

### PR DESCRIPTION
**Pull request by denniszielke**
Update cilium addon to recent version

**Reason for Change**:
feat: adds cilium support for new cilium version 

**Issue Fixed**:
Fixes #3290 

**Notes**:
Changes deployment from kube-system to cilium
yaml file is generated by the current helm from cilium repo

helm template cilium \
  --namespace cilium \
  --set global.cni.chainingMode=generic-veth \
  --set global.cni.customConf=true \
  --set global.nodeinit.enabled=true \
  --set nodeinit.azure=true \
  --set global.cni.configMap=cni-configuration \
  --set global.tunnel=disabled \
  --set global.masquerade=false

### CODE REVIEW

File: parts/k8s/addons/cilium.yaml [hunk 1]
```diff
@@ -301,71 +537,181 @@ spec:
...
           defaultMode: 420
           optional: true
           secretName: cilium-clustermesh
+        # To read the configuration from the config map
+      - configMap:
+          name: cilium-config
+        name: cilium-config-path
+      - configMap:
+          name: cni-configuration
+        name: cni-configuration
```
Comment (jackfrancis) [inline parts/k8s/addons/cilium.yaml:599]: @denniszielke we are specifying a volume to the `cni-configuration` ConfigMap, but there is no `cni-configuration` ConfigMap defined. Are we supposed to define that in this spec, or does the cilium daemonset spec depend upon the cluster already having that ConfigMap defined?

Comment (errordeveloper) [reply parts/k8s/addons/cilium.yaml:599]: Looks like this was due to `global.cni.configMap=cni-configuration` being set, so it probably means the configmap has to be present...

**Comment by errordeveloper** [comment]
Thanks a lot @denniszielke for kicking this off!!

Have you considered the [native/non-chainning mode](https://docs.cilium.io/en/v1.8/gettingstarted/k8s-install-azure/)?

File: parts/k8s/addons/cilium.yaml [hunk 2]
```diff
@@ -1,36 +1,55 @@
+---
 apiVersion: v1
-kind: Namespace
+kind: ConfigMap
 metadata:
-  labels:
```
Comment (jackfrancis) [inline parts/k8s/addons/cilium.yaml]: @denniszielke We need all these addonmanager labels in order for kube-addon-manager to do anything with this spec. Additionally, we need the Namespace resource because we are defining a new "cilium" resource.

Comment (denniszielke) [reply parts/k8s/addons/cilium.yaml]: yes, my bad - I did not merge ;)
fixed it

File: parts/k8s/addons/cilium.yaml (hunk 1 above)
Comment (denniszielke) [reply parts/k8s/addons/cilium.yaml]: added the configmap.
Do we need to change the config for kubenet ?

Comment (jackfrancis) [reply parts/k8s/addons/cilium.yaml]: I don't think there's a kubenet story here. @errordeveloper correct me if I'm wrong...

**Comment by denniszielke** [comment]
> Thanks a lot @denniszielke for kicking this off!!
> 
> Have you considered the [native/non-chainning mode](https://docs.cilium.io/en/v1.8/gettingstarted/k8s-install-azure/)?

It looks like we need to bring in the ARM auth configuration. Does this need to be explicitly configured?
Otherwise we need to use aks-engine to change the values. @jackfrancis : is this possible?

**Comment by jackfrancis** [comment]
> > Thanks a lot @denniszielke for kicking this off!!
> > Have you considered the [native/non-chainning mode](https://docs.cilium.io/en/v1.8/gettingstarted/k8s-install-azure/)?
> 
> It looks like we need to bring in the ARM auth configuration. Does this need to be explicitly configured?
> Otherwise we need to use aks-engine to change the values. @jackfrancis : is this possible?

Yes, we can inject this into the spec. If you can specify where this goes I can help with that.

**Comment by denniszielke** [comment]
> > > Thanks a lot @denniszielke for kicking this off!!
> > > Have you considered the [native/non-chainning mode](https://docs.cilium.io/en/v1.8/gettingstarted/k8s-install-azure/)?
> > 
> > 
> > It looks like we need to bring in the ARM auth configuration. Does this need to be explicitly configured?
> > Otherwise we need to use aks-engine to change the values. @jackfrancis : is this possible?
> 
> Yes, we can inject this into the spec. If you can specify where this goes I can help with that.

Ok, but this looks like a completly different cilium addon - should I add another cilium addon or can we have different version of addons?

**Comment by errordeveloper** [comment]
> Ok, but this looks like a completly different cilium addon - should I add another cilium addon or can we have different version of addons?

I think it's plausible to consider that for sure. This PR adds chaining mode for Azure CNI, but I'm not sure what's currently in tree is chained to Azure CNI, I'd guess it's more likely to be in VXLAN mode. If that's the case, should there be three add-ons or we'd be okay with a breaking change? If breaking change is okay, perhaps non-chaining mode is best to shoot for.

**Comment by errordeveloper** [comment]
> Ok, but this looks like a completly different cilium addon - should I add another cilium addon or can we have different version of addons?

I think it's plausible to consider that for sure. This PR adds chaining mode for Azure CNI, but I'm not sure what's currently in tree is chained to Azure CNI, I'd guess it's more likely to be in VXLAN mode. If that's the case, should there be three add-ons or we'd be okay with a breaking change? If breaking change is okay, perhaps non-chaining mode is best to shoot for.

/cc @tgraf

**Comment by jackfrancis** [comment]
@errordeveloper we don't need to worry about breaking the existing cilium spec, it has never tested out O.K. beyond Kubernetes 1.15 anyways. Let's assume we have a clean surface are a here for a blessed cilium 1.8 implementation on Azure.

**Comment by jackfrancis** [comment]
Just tested this locally, the current status is that the cilium daemonset pod gets stuck thusly:

```
Events:
  Type     Reason     Age                    From                                         Message
  ----     ------     ----                   ----                                         -------
  Normal   Scheduled  7m20s                  default-scheduler                            Successfully assigned cilium/cilium-5qm2f to k8s-agentpool1-41966661-vmss000001
  Normal   Pulling    7m18s                  kubelet, k8s-agentpool1-41966661-vmss000001  Pulling image "docker.io/cilium/cilium:v1.8.0"
  Normal   Pulled     7m10s                  kubelet, k8s-agentpool1-41966661-vmss000001  Successfully pulled image "docker.io/cilium/cilium:v1.8.0"
  Normal   Created    6m54s                  kubelet, k8s-agentpool1-41966661-vmss000001  Created container wait-for-node-init
  Normal   Started    6m54s                  kubelet, k8s-agentpool1-41966661-vmss000001  Started container wait-for-node-init
  Normal   Pulled     6m51s                  kubelet, k8s-agentpool1-41966661-vmss000001  Container image "docker.io/cilium/cilium:v1.8.0" already present on machine
  Normal   Started    6m50s                  kubelet, k8s-agentpool1-41966661-vmss000001  Started container clean-cilium-state
  Normal   Created    6m50s                  kubelet, k8s-agentpool1-41966661-vmss000001  Created container clean-cilium-state
  Normal   Pulled     4m24s (x3 over 6m49s)  kubelet, k8s-agentpool1-41966661-vmss000001  Container image "docker.io/cilium/cilium:v1.8.0" already present on machine
  Normal   Created    4m23s (x3 over 6m49s)  kubelet, k8s-agentpool1-41966661-vmss000001  Created container cilium-agent
  Normal   Started    4m23s (x3 over 6m49s)  kubelet, k8s-agentpool1-41966661-vmss000001  Started container cilium-agent
  Warning  BackOff    3m19s (x3 over 4m39s)  kubelet, k8s-agentpool1-41966661-vmss000001  Back-off restarting failed container
  Warning  Unhealthy  2m1s (x8 over 6m31s)   kubelet, k8s-agentpool1-41966661-vmss000001  Readiness probe failed: Get http://127.0.0.1:9876/healthz: dial tcp 127.0.0.1:9876: connect: connection refused
```

Also, the prior cilium implementation dropped a CNI config here:

`/etc/cni/net.d/05-cilium.conf`

Will we be implementing a similar CNI implementation w/ cilium 1.8?

**Comment by jackfrancis** [comment]
also cc @denniszielke I built and committed generated code

**Comment by errordeveloper** [comment]
> @errordeveloper we don't need to worry about breaking the existing cilium spec, it has never tested out O.K. beyond Kubernetes 1.15 anyways. Let's assume we have a clean surface are a here for a blessed cilium 1.8 implementation on Azure.

@jackfrancis that is what I thought was the case, thanks!

**Comment by jackfrancis** [comment]
Just did a quick smoke test of what's in this branch, the current status is that pods from the "cilium" DaemonSet are failing healthz check:

```
Warning  Unhealthy       3m49s (x6 over 6m49s)  kubelet, k8s-agentpool1-16706024-vmss000001  Readiness probe failed: Get http://127.0.0.1:9876/healthz: dial tcp 127.0.0.1:9876: connect: connection refused
```

**Comment by jackfrancis** [comment]
Just did a quick smoke test of what's in this branch, the current status is that pods from the "cilium" DaemonSet are failing healthz check:

```
Warning  Unhealthy       3m49s (x6 over 6m49s)  kubelet, k8s-agentpool1-16706024-vmss000001  Readiness probe failed: Get http://127.0.0.1:9876/healthz: dial tcp 127.0.0.1:9876: connect: connection refused
```

**Comment by denniszielke** [comment]
I think the issue might be that it requires to use networkplugin 'cilium' for cilium network policies - which is not allowed by aks-engine validation. The ds configuration is designed for azure cni
Can we change aks-e to allow azure cni + cilium np? 
Also changed the cilium pods to not run on the masters.

There seems to be something wrong with network plugin init - not sure why:
  Warning  NetworkNotReady  2m52s (x25 over 3m40s)  kubelet, k8s-agentpool1-35194500-vmss000001  network is not ready: runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized

**Comment by jackfrancis** [comment]
We definitely *don't* want to run Azure CNI in a cilium configuration (right?) to take advantage of the overlay network optimization of the container networking IP space as pods scale out.

I'll take a look today.

**Comment by jackfrancis** [comment]
So what's happening is that the CNI config is missing from `/etc/cni/net.d`. The prior cilium implementation delivered that at the filepath `/etc/cni/net.d/05-cilium.conf`, which the CSE bootstrap script expects.

tl;dr we need to deliver an equivalent CNI config under `/etc/cni/net.d`

File: parts/k8s/addons/cilium.yaml (hunk 1 above)
Comment (tgraf) [reply parts/k8s/addons/cilium.yaml]: Adding the ConfigMap as you did is fine

**Comment by tgraf** [comment]
> tl;dr we need to deliver an equivalent CNI config under `/etc/cni/net.d`

The following option will write the CNI configuration once Cilium is up:
```
write-cni-conf-when-ready: /host/etc/cni/net.d/05-cilium.conflist
```

Is something preventing Cilium from starting up?

**Comment by tgraf** [comment]
One more open question: Is something already ensuring that the BPF filesystem is being mounted? See https://docs.cilium.io/en/v1.8/concepts/kubernetes/requirements/#mounted-bpf-filesystem

It can either be done in a startup script of the node or it can be done with the node-init DaemonSet that Cilium provides.

**Comment by jackfrancis** [comment]
@tgraf, I agree, it looks like the `cilium-config` ConfigMap is delivering the correct config:

`write-cni-conf-when-ready: /host/etc/cni/net.d/05-cilium.conflist`

And the `cilium` DaemonSet has the correct volume and volumeMounts config for `/etc/cni/net.d`. I'll investigate why the CNI config isn't showing up.

**Comment by jackfrancis** [comment]
@denniszielke FYI, I rebased and made a commit to add taints/tolerations for improved bootstrap scheduling, as well as adding the addon manager label for the cilium ClusterRole.

This is the current status of the cilium daemonset. It is scheduled and runs, but I still don't see a CNI config:

```
$ ls -la /etc/cni/net.d
total 8
drwxr-xr-x 2 root root 4096 Jul  7 21:39 .
drwxr-xr-x 3 root root 4096 Jul  7 21:39 ..
azureuser@k8s-master-34786322-0:~$
```

Here are cilium pod logs from the master-scheduled pod:

```
level=info msg="  --write-cni-conf-when-ready='/host/etc/cni/net.d/05-cilium.conflist'" subsys=daemon
level=info msg="     _ _ _" subsys=daemon
level=info msg=" ___|_| |_|_ _ _____" subsys=daemon
level=info msg="|  _| | | | | |     |" subsys=daemon
level=info msg="|___|_|_|_|___|_|_|_|" subsys=daemon
level=info msg="Cilium 1.8.0 f455c7e69 2020-06-22T16:14:29+02:00 go version go1.14.4 linux/amd64" subsys=daemon
level=info msg="cilium-envoy  version: a8f292139e923b205525feb2c8a4377005904776/1.13.2/Modified/RELEASE/BoringSSL" subsys=daemon
level=info msg="clang (10.0.0) and kernel (5.3.0) versions: OK!" subsys=linux-datapath
level=info msg="linking environment: OK!" subsys=linux-datapath
level=warning msg="BPF system config check: NOT OK." error="CONFIG_BPF kernel parameter is required" subsys=linux-datapath
level=info msg="Detected mounted BPF filesystem at /sys/fs/bpf" subsys=bpf
level=info msg="Valid label prefix configuration:" subsys=labels-filter
level=info msg=" - :io.kubernetes.pod.namespace" subsys=labels-filter
level=info msg=" - :io.cilium.k8s.namespace.labels" subsys=labels-filter
level=info msg=" - :app.kubernetes.io" subsys=labels-filter
level=info msg=" - !:io.kubernetes" subsys=labels-filter
level=info msg=" - !:kubernetes.io" subsys=labels-filter
level=info msg=" - !:.*beta.kubernetes.io" subsys=labels-filter
level=info msg=" - !:k8s.io" subsys=labels-filter
level=info msg=" - !:pod-template-generation" subsys=labels-filter
level=info msg=" - !:pod-template-hash" subsys=labels-filter
level=info msg=" - !:controller-revision-hash" subsys=labels-filter
level=info msg=" - !:annotation.*" subsys=labels-filter
level=info msg=" - !:etcd_node" subsys=labels-filter
level=info msg="Using autogenerated IPv4 allocation range" subsys=node v4Prefix=10.5.0.0/16
level=info msg="Initializing daemon" subsys=daemon
level=info msg="Establishing connection to apiserver" host="https://10.0.0.1:443" subsys=k8s
level=info msg="Connected to apiserver" subsys=k8s
level=info msg="Inheriting MTU from external network interface" device=eth0 ipAddr=10.240.255.5 mtu=1500 subsys=mtu
level=info msg="Trying to auto-enable \"enable-node-port\", \"enable-external-ips\", \"enable-host-reachable-services\", \"enable-host-port\", \"enable-session-affinity\" features" subsys=daemon
level=warning msg="Session affinity for host reachable services needs kernel 5.7.0 or newer to work properly when accessed from inside cluster: the same service endpoint will be selected from all network namespaces on the host." subsys=daemon
level=info msg="Restored services from maps" failed=0 restored=0 subsys=service
level=info msg="Creating CRD (CustomResourceDefinition)..." name=CiliumNetworkPolicy/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumNetworkPolicy/v2 subsys=k8s
level=info msg="Creating CRD (CustomResourceDefinition)..." name=CiliumClusterwideNetworkPolicy/v2 subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=CiliumClusterwideNetworkPolicy/v2 subsys=k8s
level=info msg="Creating CRD (CustomResourceDefinition)..." name=v2.CiliumEndpoint subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=v2.CiliumEndpoint subsys=k8s
level=info msg="Creating CRD (CustomResourceDefinition)..." name=v2.CiliumNode subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=v2.CiliumNode subsys=k8s
level=info msg="Creating CRD (CustomResourceDefinition)..." name=v2.CiliumIdentity subsys=k8s
level=info msg="CRD (CustomResourceDefinition) is installed and up-to-date" name=v2.CiliumIdentity subsys=k8s
level=info msg="Successfully created CiliumNode resource" subsys=nodediscovery
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-34786322-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
```

The config looks like it will be writing the CNI config, but it's not present, and thus the k8s layer has no networking.

**Comment by craiglpeters** [comment]
@denniszielke is there a user who needs this? How critical is it?

**Comment by denniszielke** [comment]
> @denniszielke is there a user who needs this? How critical is it?

We have been waiting on this for a while now - so this is no longer urgent. But it would be nice if we can find a way to get it working.

**Comment by errordeveloper** [comment]
@denniszielke please do let @tgraf and myself know if there is anything we can do to unblock this!

**Comment by jackfrancis** [comment]
@errordeveloper @tgraf I just rebased this PR on top of current master and the outcome is the same as my comment above. The CNI config is not present on the node VM, and thus the cilium daemonset fails. Kubelet expects this:

```
Aug 28 16:22:53 k8s-master-17654474-0 kubelet[6846]: I0828 16:22:53.955986    6846 flags.go:33] FLAG: --cni-bin-dir="/opt/cni/bin"
Aug 28 16:22:53 k8s-master-17654474-0 kubelet[6846]: I0828 16:22:53.956110    6846 flags.go:33] FLAG: --cni-cache-dir="/var/lib/cni/cache"
Aug 28 16:22:53 k8s-master-17654474-0 kubelet[6846]: I0828 16:22:53.956240    6846 flags.go:33] FLAG: --cni-conf-dir="/etc/cni/net.d"
```

On the VM the latter two are not present:

```
azureuser@k8s-master-17654474-0:~$ ls -la /opt/cni/bin
total 84148
drwxr-xr-x 2 root root     4096 Aug 28 16:34 .
drwxr-xr-x 3 root root     4096 Aug 28 16:22 ..
-rwxr-xr-x 1 root root  4159518 May 13 19:50 bandwidth
-rwxr-xr-x 1 root root  4671647 May 13 19:50 bridge
-rwxr-xr-x 1 root root 13971456 Aug 28 16:34 cilium-cni
-rwxr-xr-x 1 root root 12124326 May 13 19:50 dhcp
-rwxr-xr-x 1 root root  5945760 May 13 19:50 firewall
-rwxr-xr-x 1 root root  3069556 May 13 19:50 flannel
-rwxr-xr-x 1 root root  4174394 May 13 19:50 host-device
-rwxr-xr-x 1 root root  3614480 May 13 19:50 host-local
-rwxr-xr-x 1 root root  4314598 May 13 19:50 ipvlan
-rwxr-xr-x 1 root root  3209463 May 13 19:50 loopback
-rwxr-xr-x 1 root root  4389622 May 13 19:50 macvlan
-rwxr-xr-x 1 root root  3939867 May 13 19:50 portmap
-rwxr-xr-x 1 root root  4590277 May 13 19:50 ptp
-rwxr-xr-x 1 root root  3392826 May 13 19:50 sbr
-rwxr-xr-x 1 root root  2885430 May 13 19:50 static
-rwxr-xr-x 1 root root  3356587 May 13 19:50 tuning
-rwxr-xr-x 1 root root  4314446 May 13 19:50 vlan
azureuser@k8s-master-17654474-0:~$ ls -la /var/lib/cni/cache
ls: cannot access '/var/lib/cni/cache': No such file or directory
azureuser@k8s-master-17654474-0:~$ ls -la /etc/cni/net.d
total 8
drwxr-xr-x 2 root root 4096 Aug 28 16:26 .
drwxr-xr-x 3 root root 4096 Aug 28 16:26 ..
```

And the cilium daemonset pod is complaining thusly:

```
level=info msg="Retrieved node information from cilium node" nodeName=k8s-master-17654474-0 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
```

**Comment by errordeveloper** [comment]
@denniszielke I think it would make sense to focus on `required IPv4 pod CIDR not present in node resource` error, I believe the issue is that `ciliumnode` (not Kubernetes `node`) doesn't have the CIRD field set. In my understanding, and I could be wrong, it's the the `cilium-operator` that is meant to set it, and usually it's be based on `global.nativeRoutingCIDR`.

Maybe you could try using `--set global.nativeRoutingCIDR=<defaultCIDR>` and see if that fixes this issue? Once we established that, we can see how this CIRD can be parametrised at runtime.

**Comment by jackfrancis** [comment]
@errordeveloper is the `--set global.nativeRoutingCIDR=<defaultCIDR>` a CLI option for the `cilium-operator-generic` command in the `cilium-operator` deployment container?

**Comment by jackfrancis** [comment]
For the record I don't think the cilium-operator component is at fault here, it's not even able to be scheduled due to no nodes being Ready:

```
  Warning  FailedScheduling  23m         0/3 nodes are available: 1 node(s) had taint {node-role.kubernetes.io/master: true}, that the pod didn't tolerate, 2 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate.
```

**Comment by jackfrancis** [comment]
So, I was able to move things further along by changing the config for `k8s-require-ipv4-pod-cidr` to `"false"`, but it seems that the `azure-vnet` binary is not being dropped by the CNI daemonset:

```
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]: W0901 20:12:00.478056    1190 cni.go:204] Error validating CNI config list {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   "cniVersion": "0.3.0",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   "name": "azure",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   "plugins": [
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "type": "azure-vnet",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "mode": "transparent",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "bridge": "azure0",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "ipam": {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:          "type": "azure-vnet-ipam"
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:        }
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     },
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "type": "portmap",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "capabilities": {"portMappings": true},
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "snat": true
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     },
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:        "name": "cilium",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:        "type": "cilium-cni"
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     }
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   ]
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]: }: [failed to find plugin "azure-vnet" in path [/opt/cni/bin]]
```

**Comment by errordeveloper** [comment]
I was referring to helm value.

**Comment by errordeveloper** [comment]
On Tue, 1 Sep 2020, 9:13 pm Jack Francis, <notifications@github.com> wrote:

> So, I was able to move things further along by changing the config for
> k8s-require-ipv4-pod-cidr to "false", but it seems that the azure-vnet
> binary is not being dropped by the CNI daemonset:
>
Do you mean Azure CNI daemonset?

File: parts/k8s/addons/cilium.yaml [hunk 3]
```diff
@@ -1,69 +1,457 @@
...
+  masquerade: "false"
+  enable-bpf-masquerade: "true"
+  enable-xt-socket-fallback: "true"
+  install-iptables-rules: "true"
+  auto-direct-node-routes: "false"
+  kube-proxy-replacement:  "probe"
+  node-port-bind-protection: "true"
+  enable-auto-protect-node-port-range: "true"
+  enable-session-affinity: "true"
+  k8s-require-ipv4-pod-cidr: "true"
```
Comment (jackfrancis) [inline parts/k8s/addons/cilium.yaml:196]: @errordeveloper I mean that if you change `k8s-require-ipv4-pod-cidr` to "false", that the CNI configuration is properly delivered by the cilium daemonset. But then kubelet complains thusly:

```
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]: W0901 20:12:00.478056    1190 cni.go:204] Error validating CNI config list {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   "cniVersion": "0.3.0",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   "name": "azure",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   "plugins": [
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "type": "azure-vnet",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "mode": "transparent",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "bridge": "azure0",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "ipam": {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:          "type": "azure-vnet-ipam"
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:        }
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     },
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "type": "portmap",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "capabilities": {"portMappings": true},
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:       "snat": true
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     },
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     {
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:        "name": "cilium",
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:        "type": "cilium-cni"
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:     }
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]:   ]
Sep 01 20:12:00 k8s-master-10101788-0 kubelet[1190]: }: [failed to find plugin "azure-vnet" in path [/opt/cni/bin]]
```

**Comment by errordeveloper** [comment]
> So, I was able to move things further along by changing the config for `k8s-require-ipv4-pod-cidr` to `"false"`, but it seems that the `azure-vnet` binary is not being dropped by the CNI daemonset:

Are you referring to Azure CNI daemonset?

**Comment by errordeveloper** [comment]
> @errordeveloper is the `--set global.nativeRoutingCIDR=<defaultCIDR>` a CLI option for the `cilium-operator-generic` command in the `cilium-operator` deployment container?

I was suggesting this to be added to helm flags (see PR description).

**Comment by jackfrancis** [comment]
@errordeveloper No, I'm referring to changing the cilium YAML spec that's defined in this PR.

**Comment by jackfrancis** [comment]
@errordeveloper I understand your comment now. Yeah, I tried adding the equivalent configuration directly to the spec's ConfigMap:

`native-routing-cidr: "10.244.0.0/16"`

(`10.244.0.0/16` is the default cluster CIDR for aks-engine-created clusters. That's the value, for example, passed to the `kube-controller-manager` component for its `--cluster-cidr` configuration.)

Doing *just* that gets this error still:

```
level=info msg="Retrieved node information from cilium node" nodeName=k8s-agentpool1-12337648-vmss000001 subsys=k8s
level=warning msg="Waiting for k8s node information" error="required IPv4 pod CIDR not present in node resource" subsys=k8s
```

That error points to this line in the initialization code:

https://github.com/cilium/cilium/blob/v1.8.0/pkg/k8s/init.go#L120

Basically, what that code tells me is that the "require ipv4 CIDR" configuration is set to true (indeed it is in the YAML spec), but that the node object has no ipv4 CIDR value. I don't know how to fix that. Is there another configuration value we need to set so that the node object during initiatlization has the correct CIDR value?

File: parts/k8s/addons/cilium.yaml (hunk 3 above)
Comment (errordeveloper) [reply parts/k8s/addons/cilium.yaml:196]: Is `azure-vnet` installation somehow disabled by this configuration? What would it take ensure it's installed? Or is it in  a different path?

**Comment by errordeveloper** [comment]
Perhaps, if we cannot get chaining mode to work, it would make sense to try native Azure IPAM mode first, and revisit the chaining mode later, if chaining mode is still needed?

**Comment by jackfrancis** [comment]
Hi @errordeveloper by "native Azure IPAM" do we mean Azure CNI? (we don't want that, we want the benefits of reduce IP address overhead offered by the cilium overlay"

**Comment by errordeveloper** [comment]
@jackfrancis no, native Azure IPAM mode (see [docs](https://docs.cilium.io/en/v1.8/gettingstarted/k8s-install-azure/)) will avoid Azure CNI, but it will used native IP addresses and allocate them using Azure APIs directly from `cilium-operator`, so it will not overcome address space limitations. If the intention of this PR is to deploy Cilium in overlay mode, as that's not what the PR does as it stands, as it uses `global.cni.chainingMode=generic-veth` and `global.tunnel=disabled` (this is the options that actually disables overlay). I think it's probably a good idea to start fresh in a new PR, and close this one :)
